Telemetry: Add custom label support to metrics

kkfernandezTT opened the pull request:
### Ticket
Addresses reviewer feedback on PR #32120 and enables process metrics implementation.

### Problem description
The telemetry system only supported hierarchical path-based labels (hostname, tray, chip, channel) which are derived from the metric's position in the system topology. This made it impossible to add runtime contextual metadata like user, process name, or physical port information without:
1. Hacky path parsing (string manipulation with magic numbers)
2. Encoding data in paths (ugly, fragile)
3. Creating separate StringMetrics (wasteful, not Prometheus-idiomatic)

Reviewer feedback on PR #32120 requested proper attribute support on the Metric base class instead of path parsing hacks.

### What's changed
Implemented a comprehensive custom label system following Prometheus best practices:

**Core Infrastructure:**
- Extended `Metric` base class with `custom_labels_` map and API (`set_label()`, `set_labels()`, `labels()`)
- Added unified `metric_labels` map to `TelemetrySnapshot` (single map for all metric types)
- Updated `telemetry_collector` to populate labels in snapshots with granular delta transmission
- Modified `prom_formatter` to merge custom labels with path-derived labels

**Quality & Correctness:**
- Prometheus label key validation with ASCII-only checks (locale-independent)
- Validation works in release builds (log_error, not just assertions)
- Label conflict detection and rejection during merge
- Orphaned label prevention (tracks successfully merged metrics)
- Efficient change tracking (only send labels when changed)
- Variadic template for extensible path validation

**API Design:**
- Uses `std::optional<std::reference_wrapper<>>` instead of raw pointers
- Helper functions (`is_ascii_alpha`, `is_ascii_alnum`, `path_exists_in_any_map`)
- Comprehensive inline documentation with usage examples
- Backward compatible JSON serialization

**Testing:**
- 12 comprehensive unit tests covering all functionality
- Tests for validation, change tracking, merge, conflicts, orphaned labels

**Impact:**
Enables process metrics to store identity information as labels:
```cpp
metric->set_label("user", "alice");
metric->set_label("process", "python3");
// Output: cpu{hostname="...",device="0",user="alice",process="python3"} 85.2
```

Enables Ethernet metrics to store physical topology as labels (replaces path parsing):
```cpp
metric->set_label("port_type", "QSFP_DD");
metric->set_label("port_id", "3");
```

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/docs/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml) CI passes (if applicable)
- [ ] [Galaxy demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) CI passes, if applicable, because of current Llama work
- [ ] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes (if applicable, since this is run on push to main)
- [ ] (For models and ops writers) [T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [x] New/Existing tests provide coverage for changes

kkfernandezTT committed:
Add custom label support to telemetry metrics

Implements a flexible label system for Prometheus metrics that allows metrics
to carry arbitrary key-value metadata beyond the hierarchical path structure.

Core changes:
- Add custom_labels_ map to Metric base class with set_label() API
- Add unified metric_labels map to TelemetrySnapshot for all metric types
- Update telemetry_collector to populate labels in snapshots
- Modify prom_formatter to merge custom labels with path-derived labels
- Use std::optional instead of raw pointers for optional parameters
- Add Prometheus label key validation (enforces [a-zA-Z_][a-zA-Z0-9_]*)
- Add label conflict detection in merge validation
- Cache label references to avoid redundant calls

Benefits:
- Enables rich contextual metadata on metrics (user, process, etc.)
- Prometheus-native approach (labels in same {} block as path labels)
- Efficient delta transmission with change tracking
- Backward compatible JSON serialization
- Single unified label map simplifies code (~70% reduction vs per-type maps)

Example usage:
  metric->set_label("user", "alice");
  metric->set_label("process", "python3");

Prometheus output:
  cpu{hostname="...",device="0",user="alice",process="python3"} 85.2

kkfernandezTT committed:
Fix locale-dependent validation and label merge semantics

- Use ASCII-only checks instead of std::isalpha/isalnum (locale-independent)
- Add helper functions is_ascii_alpha() and is_ascii_alnum()
- Make validation work in release builds (log_error instead of TT_ASSERT)
- Track successfully merged paths to prevent orphaned labels from rejected metrics
- Reject label merges on conflict instead of last-write-wins
- Replace cctype include with tt-logger for runtime validation

kkfernandezTT committed:
Add comprehensive unit tests for custom label functionality

Tests cover:
- Basic label setting and retrieval
- Multiple labels per metric
- Label updates and change tracking
- Bulk label updates with set_labels()
- Label validation (valid/invalid keys)
- Snapshot label storage and merge
- Orphaned label rejection
- Label conflict detection and rejection
- Independent value/label change tracking
- Empty label handling

kkfernandezTT committed:
Fix label adoption from rejected metrics

Prevent labels from a rejected metric (due to type conflict) from being
incorrectly applied to an existing metric with the same path.

The bug: When 'other' sends a metric that conflicts with base's type,
the metric is rejected but its labels were still merged onto the base
metric of a different type.

Fix: Check if 'other' sent a metric value for the path. If yes and it
wasn't successfully merged, reject its labels too (they belong to the
rejected metric, not the existing one).

This fixes the LabelsOnlyMergedForSuccessfulMetrics test.

btrzynadlowski-tt commented on tt_telemetry/include/telemetry/metric.hpp:
```diff
@@ -81,7 +181,9 @@ class Metric {
     bool changed_since_transmission_ = false;
+    bool labels_changed_since_transmission_ = false;
     uint64_t timestamp_ = 0;  // Unix timestamp in milliseconds, 0 = never set
+    std::unordered_map<std::string, std::string> custom_labels_;
```
This will end up being the only thing we need to expose via a const getter. Ideally, the label map would be constructed in the constructor and remain immutable thereafter.

btrzynadlowski-tt replied:
Additionally, I'm on the fence as to whether the base Metric class even needs to define a map like this. It definitely needs the getter (with a default empty map). But for example, look at Ethernet metrics. We would 1) rarely read the attribute map and 2) already store the data that would go inside of them (asic location, Ethernet channel, tray number, etc.) We could construct a map in-place and return it when asked, which will only happen the first time a snap is sent over to the web thread.

kkfernandezTT committed:
Make labels immutable

kkfernandezTT committed:
Fix code review issues: add missing include, handle localtime_r failure, prevent log spam

- Add missing #include <unordered_set> to telemetry_snapshot.hpp
- Check localtime_r return value with fallback to gmtime_r
- Prevent log spam in prom_formatter with per-conflict suppression
- Remove unreachable dead code in get_cluster_wide_telemetry_path
- Remove unused tt-logger include from metric.hpp

kkfernandezTT committed:
Fix compile error with wrong log level
